src/analysis/retail/evoker/preservation/normalizers/CastLinkNormalizer.ts

```diff
@@ -0,0 +1,109 @@
+  */
+const EVENT_LINKS: EventLink[] = [
+  // link Echo apply to its CastEvent
+  {
+    linkRelation: FROM_HARDCAST,
+    linkingEventId: [TALENTS_EVOKER.ECHO_TALENT.id],
+    linkingEventType: [EventType.ApplyBuff, EventType.RefreshBuff],
+    referencedEventId: TALENTS_EVOKER.ECHO_TALENT.id,
+    referencedEventType: [EventType.Cast],
+    forwardBufferMs: CAST_BUFFER_MS,
+    backwardBufferMs: CAST_BUFFER_MS,
+  },
+  //link hardcast echo removal to hot application
+  {
+    linkRelation: ECHO,
+    linkingEventId: [SPELLS.REVERSION_ECHO.id, SPELLS.DREAM_BREATH_ECHO.id],
+    linkingEventType: [EventType.ApplyBuff, EventType.RefreshBuff],
+    referencedEventId: TALENTS_EVOKER.ECHO_TALENT.id,
+    referencedEventType: [EventType.RemoveBuff],
+    forwardBufferMs: CAST_BUFFER_MS,
+    backwardBufferMs: CAST_BUFFER_MS,
+    additionalCondition(referencedEvent) {
+      return !HasRelatedEvent(referencedEvent, FROM_TEMPORAL_ANOMALY);
+    },
+  },
+  //link echo apply to the Temporal Anomaly shield application
+  {
+    linkRelation: FROM_TEMPORAL_ANOMALY,
+    linkingEventId: [TALENTS_EVOKER.ECHO_TALENT.id],
+    linkingEventType: [EventType.ApplyBuff, EventType.RefreshBuff],
+    referencedEventId: SPELLS.TEMPORAL_ANOMALY_SHIELD.id,
+    referencedEventType: [EventType.ApplyBuff, EventType.RefreshBuff],
+    forwardBufferMs: CAST_BUFFER_MS,
+    backwardBufferMs: CAST_BUFFER_MS,
+    isActive(c) {
+      return c.hasTalent(TALENTS_EVOKER.TEMPORAL_ANOMALY_TALENT);
+    },
+  },
+  //link TA echo removal to hot application
+  {
+    linkRelation: ECHO_TEMPORAL_ANOMALY,
+    linkingEventId: [SPELLS.REVERSION_ECHO.id, SPELLS.DREAM_BREATH_ECHO.id],
+    linkingEventType: [EventType.ApplyBuff, EventType.RefreshBuff],
+    referencedEventId: TALENTS_EVOKER.ECHO_TALENT.id,
+    referencedEventType: [EventType.RemoveBuff],
+    forwardBufferMs: CAST_BUFFER_MS,
+    backwardBufferMs: CAST_BUFFER_MS,
+    additionalCondition(referencedEvent) {
+      return !HasRelatedEvent(referencedEvent, FROM_HARDCAST);
+    },
+    isActive(c) {
+      return c.hasTalent(TALENTS_EVOKER.TEMPORAL_ANOMALY_TALENT);
+    },
+  },
+];
+
+/**
+ * When a spell is cast on a target, the ordering of the Cast and ApplyBuff/RefreshBuff/(direct)Heal
+ * can be semi-arbitrary, making analysis difficult.
+ *
+ * This normalizer adds a _linkedEvent to the ApplyBuff/RefreshBuff/RemoveBuff linking back to the Cast event
+ * that caused it (if one can be found).
+ *
+ * This normalizer adds links for Echo and Temporal Anomaly
+ */
+class CastLinkNormalizer extends EventLinkNormalizer {
+  constructor(options: Options) {
+    super(options, EVENT_LINKS);
+  }
+}
+
+/** Returns true iff the given buff application or heal can be matched back to a hardcast */
+export function isFromHardcast(event: AbilityEvent<any>): boolean {
+  if (HasRelatedEvent(event, ECHO_TEMPORAL_ANOMALY)) {
+    return false;
+  }
+  if (HasRelatedEvent(event, ECHO)) {
+    return true;
+  }
+  return false;
+}
```

Comment:
Layout of this function is a bit obtuse, can just do `return HasRelatedEvent(event, ECHO)`. Only way I see this being an issue is if you are linking echo TA and echo and if thats the case, then you should adjust the linking.